### PR TITLE
Removed functions code && supported functions plugin

### PR DESCRIPTION
**Pull request by klakhov**
### Checklist

- [ ] I submit my changes into the `develop` branch
- [ ] I have added a description of my changes into the [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary
  ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [ ] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

### CODE REVIEW

File: cvat-core/src/server-proxy.ts [hunk 1]
```diff
@@ -483,6 +482,7 @@ async function healthCheck(
     progressCallback: (status: string) => void,
     attempt = 0,
 ): Promise<void> {
+    return true;
```
Comment (SpecLad) [inline cvat-core/src/server-proxy.ts:485]: This looks like it was unintentional.

File: cvat-core/src/api-implementation.ts [hunk 2]
```diff
@@ -117,6 +119,11 @@ export default function implementAPI(cvat) {
         return result;
     };
 
+    cvat.server.requestAll.implementation = async (url) => {
+        const result = await serverProxy.server.requestAll(url);
+        return result;
+    };
+
```
Comment (bsekachev) [inline cvat-core/src/api-implementation.ts:126]: I believe we do not need extra method for this action. 
Better to add some arguments to the existing method

File: cvat-core/src/lambda-manager.ts [hunk 3]
```diff
@@ -14,7 +14,12 @@ export interface ModelProvider {
     attributes: Record<string, string>;
 }
 
-interface ModelProxy {
+export enum ModelRequestStatus {
+    QUEUED = 'queued',
+    STARTED = 'started',
+}
```
Comment (bsekachev) [inline cvat-core/src/lambda-manager.ts:20]: We should find a way to remove all the code related with models from the core application

File: cvat-core/src/ml-model.ts [hunk 4]
```diff
@@ -157,27 +151,7 @@ Object.defineProperties(MLModel.prototype.getPreview, {
...
-                                resolve(reader.result);
-                            };
-                            reader.readAsDataURL(result);
-                        }
-                    })
-                    .catch((error) => {
-                        reject(error);
-                    });
-            });
+            return '';
```
Comment (bsekachev) [inline cvat-core/src/ml-model.ts:154]: Do we need this method now?

File: cvat-ui/src/reducers/plugins-reducer.ts [hunk 5]
```diff
@@ -25,6 +24,17 @@ const defaultState: PluginsState = {
         loginPage: {
             loginForm: [],
         },
+        modelsPage: {
+            topBar: {
+                items: [],
+            },
+            deployedModelItem: {
```
Comment (bsekachev) [inline cvat-ui/src/reducers/plugins-reducer.ts:31]:
```suggestion
            modelItem: {
```

File: cvat-ui/src/reducers/plugins-reducer.ts [hunk 6]
```diff
@@ -25,6 +24,17 @@ const defaultState: PluginsState = {
...
+                menu: {
+                    items: [],
+                },
+                icon: [],
```
Comment (bsekachev) [inline cvat-ui/src/reducers/plugins-reducer.ts:35]: What do you mean here? "Icon" does not sound like a place where we can add some components

File: cvat-ui/src/reducers/index.ts [hunk 7]
```diff
@@ -263,6 +263,11 @@ export type PluginsList = {
     [name in SupportedPlugins]: boolean;
 };
 
+export interface PluginsRegistrar {
+    modelsPlugin: () => void;
```
Comment (bsekachev) [inline cvat-ui/src/reducers/index.ts:267]: Does core app know about models plugin?

File: cvat-core/src/ml-model.ts (hunk 4 above)
Comment (klakhov) [reply cvat-core/src/ml-model.ts:154]: Yes, we still need to work with empty previews for Lambda functions

File: cvat-ui/src/reducers/plugins-reducer.ts (hunk 6 above)
Comment (klakhov) [reply cvat-ui/src/reducers/plugins-reducer.ts:35]: Why? Here I mean plugins in ModelItem icon component which represents its provider

File: cvat-ui/src/reducers/index.ts (hunk 7 above)
Comment (klakhov) [reply cvat-ui/src/reducers/index.ts:267]: No, it's core app plugin that helps cvat-ui to react on LambdaManager model list changes. Core app doesnt know when and how this list changes. Thats how lambda functions and functions from external providers are getting mixed in one list

File: cvat-ui/src/reducers/plugins-reducer.ts (hunk 6 above)
Comment (bsekachev) [reply cvat-ui/src/reducers/plugins-reducer.ts:35]: I understand what is ``ModelItem``, but ``icon`` looks like thing what we need to integrate, but not the place where we need to integrate

File: cvat-core/src/ml-model.ts (hunk 4 above)
Comment (bsekachev) [reply cvat-core/src/ml-model.ts:154]: Embedded functions don't have a preview. So, why we need this method?

Comment (bsekachev) [reply cvat-core/src/ml-model.ts:154]: It looks strange that it just returns empty string.

File: cvat-ui/src/reducers/index.ts (hunk 7 above)
Comment (bsekachev) [reply cvat-ui/src/reducers/index.ts:267]: Need to discuss it

File: cvat-ui/src/reducers/plugins-reducer.ts (hunk 6 above)
Comment (klakhov) [reply cvat-ui/src/reducers/plugins-reducer.ts]: Changed to `topBar` integration

File: cvat-core/src/api-implementation.ts [hunk 8]
```diff
@@ -39,12 +39,14 @@ export default function implementAPI(cvat) {
     cvat.plugins.register.implementation = PluginRegistry.register.bind(cvat);
 
     cvat.lambda.list.implementation = lambdaManager.list.bind(lambdaManager);
+    cvat.lambda.updateModelList.implementation = lambdaManager.updateModelList.bind(lambdaManager);
+    cvat.lambda.setModelProxyMap.implementation = lambdaManager.setModelProxyMap.bind(lambdaManager);
+    cvat.lambda.getCachedList.implementation = lambdaManager.getCachedList.bind(lambdaManager);
```
Comment (bsekachev) [inline cvat-core/src/api-implementation.ts:44]: That should not exist in the core code

Comment (bsekachev) [reply cvat-core/src/api-implementation.ts:44]: Let's discuss in a call

Comment (klakhov) [reply cvat-core/src/api-implementation.ts]: Removed

File: cvat-core/src/ml-model.ts [hunk 9]
```diff
@@ -127,15 +125,11 @@ Object.defineProperties(MLModel.prototype.save, {
...
-                api_key: this.serialized.api_key,
-            };
-
-            const model = await serverProxy.functions.create(modelData);
-            return new MLModel(model);
+        value: async function implementation(): Promise<MLModel> {
+            if (this.provider === ModelProviders.CVAT) {
+                throw Error('Saving built-in models is not available.');
+            }
+            return this;
```
[the remark this answers is not in the document]
Comment (klakhov) [reply cvat-core/src/ml-model.ts]: removed

File: cvat-core/src/lambda-manager.ts [hunk 10]
```diff
@@ -6,44 +6,37 @@
...
 }
 
-interface ModelProxy {
-    run: (body: any) => Promise<any>;
-    call: (modelID: string | number, body: any) => Promise<any>;
-    status: (requestID: string) => Promise<any>;
-    cancel: (requestID: string) => Promise<any>;
+export enum ModelRequestStatus {
+    QUEUED = 'queued',
+    STARTED = 'started',
```
Comment (bsekachev) [inline cvat-core/src/lambda-manager.ts:19]: Why can't we use RQStatus?

File: cvat-core/src/ml-model.ts [hunk 11]
```diff
@@ -107,60 +105,18 @@ export default class MLModel {
...
 Object.defineProperties(MLModel.prototype.preview, {
     implementation: {
         writable: false,
         enumerable: false,
-        value: async function implementation(this: MLModel): Promise<string> {
-            if (this.provider === ModelProviders.CVAT) return '';
-            const preview = await serverProxy.functions.getPreview(this.id);
-            if (!preview) return '';
-            return decodePreview(preview);
+        value: async function implementation(): Promise<string | ArrayBuffer | null> {
```
Comment (bsekachev) [inline cvat-core/src/ml-model.ts:118]: As far as I remember preview methods may return only string or null

File: cvat-core/src/server-proxy.ts [hunk 12]
```diff
@@ -569,9 +569,24 @@ async function healthCheck(
...
-async function serverRequest(url: string, data: object): Promise<any> {
-    try {
-        const res = await Axios(url, data);
+export interface ServerRequestConfig {
+    fetchAll: boolean,
+}
+const defaultRequestConfig = {
+    fetchAll: false,
+};
+async function serverRequest(
```
Comment (bsekachev) [inline cvat-core/src/server-proxy.ts:578]:
```suggestion
}

const defaultRequestConfig = {
    fetchAll: false,
};

async function serverRequest(
```

File: cvat-ui/src/components/models-page/model-provider-icon.tsx [hunk 13]
```diff
@@ -3,27 +3,24 @@
 // SPDX-License-Identifier: MIT
 
 import React from 'react';
-import { ModelProvider } from 'cvat-core-wrapper';
+import { usePlugins } from 'utils/hooks';
```
Comment (bsekachev) [inline cvat-ui/src/components/models-page/model-provider-icon.tsx:6]: The component itself better to be moved to a plugin
Or icon should be added for system models

File: cvat-core/src/lambda-manager.ts (hunk 10 above)
Comment (klakhov) [reply cvat-core/src/lambda-manager.ts]: Done

File: cvat-ui/src/components/models-page/model-provider-icon.tsx (hunk 13 above)
Comment (klakhov) [reply cvat-ui/src/components/models-page/model-provider-icon.tsx]: This component wasn't used(already moved to plugin), deleted it.

**Review by bsekachev — APPROVED** [review]
LGTM, but postpone merge, because I need first to look at the second part

**Pull request merged**